pwndbg/commands/chain.py

```diff
@@ -0,0 +1,394 @@
+def chain_cmd(path, next, sentinel, inner_name, field_name) -> None:
+    # Have GDB parse the path for us and check if it's valid.
+    try:
+        first = gdb.parse_and_eval(path)
+    except gdb.error as e:
+        print(message.error(f"{e}"))
+        return
+
+    if first.is_optimized_out:
+        print(message.error(f"{path} has been optimized out"))
+        return
+
+    # We suport being passed either a pointer to the first structure or the
+    # structure itself, for the sake of convenience. But we don't bother with
+    # chains of pointers. Additionally, we pick the correct separator to use
+    # for the error mesages.
+    sep = "."
+    deref = ""
+    if first.type.code == gdb.TYPE_CODE_PTR:
+        sep = "->"
+        deref = "*"
+        try:
+            first = first.dereference()
+        except gdb.error as e:
+            print(message.error(f"Pointer at {path} could not be dereferenced: {e}"))
+            return
+
+    if first.type.code == gdb.TYPE_CODE_PTR:
+        print(message.error(f"{path} is not a value or a single pointer to one"))
+        return
+
+    if first.address is None:
+        print(message.error(f"{deref}{path} is not addressable"))
+
+    if first.is_optimized_out:
+        print(message.error(f"{deref}{path} has been optimized out"))
+
+    # If there is an inner element we have to use, find it.
+    inner = None
+    inner_sep = ""
+    if inner_name is not None:
+        try:
+            inner = first[inner_name]
+            inner_sep = "->"
+        except gdb.error as e:
+            print(message.error(f"Cannot find component {inner_name} in {path}: {e}"))
+            return
+        if inner.is_optimized_out:
+            print(message.error(f"{path}{sep}{inner_name} has been optimized out"))
+            return
+
+    # Resolve the pointer to the next structure, wherever it may be, and make
+    # sure that we can use it to traverse the chain.
+    next_ptr_loc = first
+    next_ptr_name = next
+    try:
+        if inner is None:
+            next_ptr = first[next]
+        else:
+            next_ptr = inner[next]
+            next_ptr_loc = inner
+            next_ptr_name = f"{inner_name}.{next}"
+    except gdb.error as e:
+        print(message.error(f"Cannot find component {next_ptr_name} in {path}: {e}"))
+        return
+
+    if next_ptr.is_optimized_out:
+        print(message.error(f"{path}{sep}{next_ptr_name} has been optimized out"))
+        return
+    if next_ptr.type.code != gdb.TYPE_CODE_PTR:
+        print(message.error(f"{path}{sep}{next_ptr_name} is not a pointer"))
+        return
+
+    # Helper functions to resolve the offsets of fields inside a structure.
+    def bit_offset_of_field(struct, field_name, inner_name=None):
+        offset = None
+        for field in struct.type.fields():
+            if field.name == field_name:
+                offset = field.bitpos
+        return offset
+
+    def get_byte_offset(bit_offset):
```

Comment:
Can we move the two helper functions out of the command function?

Defining a nested function in Python (CPython) creates this function dynamically on each execution of the outer function. This can be observed by inspecting the bytecode generated for an outer function:

```py
 In [4]: def foo():
   ...:     def bar():
   ...:         pass
   ...: 

In [5]: dis.dis(foo)
  2           0 LOAD_CONST               1 (<code object bar at 0x7f605d76a550, file "<ipython-input-4-3215c711badf>", line 2>)
              2 LOAD_CONST               2 ('foo.<locals>.bar')
              4 MAKE_FUNCTION            0
              6 STORE_FAST               0 (bar)
              8 LOAD_CONST               0 (None)
             10 RETURN_VALUE

Disassembly of <code object bar at 0x7f605d76a550, file "<ipython-input-4-3215c711badf>", line 2>:
  3           0 LOAD_CONST               0 (None)
              2 RETURN_VALUE
```

Reply:
Yes, I've moved them now. Guess I'm so used to this being a non-issue in Rust that the possibility didn't even cross my mind.